Move soft keyword definitions outside the parser

tgodzik opened the pull request:
Based on https://github.com/scalameta/scalafmt/pull/2284#discussion_r538886558

This way they can be reused inside scalafmt.

olafurpg commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:
```diff
@@ -0,0 +1,87 @@
+    case _ => false
+  }
+
+  def isSoftKw(token: Token, name: String): Boolean = isIdentAnd(token, _ == name)
```
Can this be `private`?

kitbellew replied:
also, could add dialect check as well, so it's always in the right order of checking:
```
  private matches(token: Token, name: String, isEnabled: Dialect => Boolean): Boolean =
    isEnabled(dialect) && isIdentAnd(token, _ == name)
```
(in this class, everything is about soft keywords, so i also renamed the method)

tgodzik replied:
Moved it actually to ScalametParserObject and is now reused in SoftKeywords and ScalametaParser.

tgodzik replied:
Ach no, not that one :sweat_smile: this one is renamed to `matches` and made private.

olafurpg commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:
```diff
@@ -0,0 +1,87 @@
+
+class SoftKeywords(dialect: Dialect) {
+
+  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
```
Make private?

kitbellew replied:
actually, this duplicates the method in `ScalametaParser` above. should we do this instead:

```
private[parsers] object ScalametaParser {
  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
    case Ident(value) => pred(value.stripPrefix("`").stripSuffix("`"))
    case _ => false
  }
}
```
and then import it in both classes?

tgodzik replied:
Moved it actually to ScalametParserObject and is now reused in SoftKeywords and ScalametaParser.

kitbellew commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -837,77 +840,6 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
-      isSoftKw(token, name) && dialect.allowInfixMods
-    }
-  }
-
```
why not extract all the other classifiers, too? lots of them are obviously useful to like whitespace, trivia etc.

tgodzik replied:
Most of it is internal to parser, so to be honest I am not sure what would make sense to extract. I will start with the soft keywords now, but let me know if you find anything that might be particularily useful in scalafmt.

olafurpg commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:
```diff
@@ -0,0 +1,87 @@
+  def isSoftKw(token: Token, name: String): Boolean = isIdentAnd(token, _ == name)
+
+  @classifier
+  trait SkAs {
```
Nitpick: I'm not a fan of the "Sk" prefix. Instead of using wildcard imports, how do you like this syntax?
```suggestion
class SoftKeywords {
  trait KwAs { ... }
}

val soft = new SoftKeywords()
x match {
  case soft.KwAs
```

tgodzik replied:
Looks good, I will change it!

tgodzik replied:
Reafctored!

olafurpg commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:
```diff
@@ -0,0 +1,87 @@
+  trait SkAs {
+    val name = "as"
+    def unapply(token: Token): Boolean = {
+      isSoftKw(token, name) && dialect.allowAsPatternBinding
```
Nit: can we move `dialect.allowAsPatternBinding` before the `isSoftKw`? The `isSoftKw` test is a lot more expensive and will be `false` in most cases

tgodzik replied:
Sure! Fun fact, this was added with M1 and will actually be removed with M3.

tgodzik replied:
Used the advice from @kitbellew and now we have a matches method that first always checks if a particular dialect feature is supported.

kitbellew commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:
```diff
@@ -0,0 +1,87 @@
+class SoftKeywords(dialect: Dialect) {
+
+  def isIdentAnd(token: Token, pred: String => Boolean): Boolean = token match {
+    case Ident(value) if pred(value.stripPrefix("`").stripSuffix("`")) => true
```
very optional:
```
  case Ident(value) => pred(value.stripPrefix("`").stripSuffix("`"))
```

tgodzik replied:
Fixed, looks much nicer.

kitbellew commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -837,77 +840,6 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
       (isCaseClassOrObject || (token.is[KwCase] && token.next.is[Ident] && dialect.allowEnums))
   }
 
-  @classifier
```
(stupid github only allows commenting on red/green lines) should we remove `isSoftKw` above?

tgodzik replied:
Ach right! Removed it!

kitbellew commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SoftKeywords.scala:
```diff
@@ -0,0 +1,87 @@
+import scala.meta.tokens.Token.Ident
+import scala.meta.internal.classifiers.classifier
+
+class SoftKeywords(dialect: Dialect) {
```
do we need to wrap this in an instance? wouldn't this work, with implicit:
```
  object SoftKw {
    @classifier
    trait Using {
      def unapply(token: Token)(implicit dialect: Dialect): Boolean = {
        dialect.allowGivenUsing && isSoftKw(token, "using")
      }
    }
  }

  // somewhere else
  token.is[SoftKw.Using]
```

tgodzik replied:
Tried it and it does not recognize it as an unapply method :/

kitbellew reviewed (changes requested):
what started as a simple refactor...

tgodzik commented:
> also, could add dialect check as well, so it's always in the right order of checking:
>
>  private matches(token: Token, name: String, isEnabled: Dialect => Boolean): Boolean =
>    isEnabled(dialect) && isIdentAnd(token, _ == name)
> (in this class, everything is about soft keywords, so i also renamed the method)

Fixed it, great suggestion! 

> actually, this duplicates the method in ScalametaParser above. should we do this instead:

Moved it, thanks! @kitbellew

kitbellew commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -36,6 +36,8 @@ import scala.util.Failure
 class ScalametaParser(input: Input, dialect: Dialect) { parser =>
   require(Set("", EOL).contains(dialect.toplevelSeparator))
   implicit val currentDialect: Dialect = dialect
+  val soft = new SoftKeywords(currentDialect)
```
somewhat optional: 1. `private` 2. use `dialect` directly, and not the intermediate `currentDialect`?

also, for a completely different, unrelated change: we could change the class to
```
class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
   ...
```
and remove `currentDialect`.

tgodzik replied:
Good idea! Changed!

kitbellew commented:
@tgodzik should we release and use in `scalafmt`?

tgodzik commented:
Sure, makes sense. We have two more PRs lined up, but it shouldn't really influence scalafmt. I will run it now.
